tao_compiler/mlir/xla/ral/context/common_context_impl_mkldnn.cc

```diff
@@ -556,6 +557,74 @@ struct OnednnGemmState : public Context::Resource {
   std::unordered_map<opaque_t, std::vector<ideep::tensor>> packed_weight_cache;
 };
 
+template <class T>
+inline void hash_combine(std::size_t& seed, const T& v) {
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+}
+
+struct OnednnACLGemmKey {
+  int m = -1;
+  int n = -1;
+  int k = -1;
+  int batch = 1;
+  bool transpose_a = false;
+  bool transpose_b = false;
+  opaque_t const_weight_ptr = nullptr;
+  // We need this since the matmul primitive is not thead safe.
+  // To enable large parallelism, we cache the primitive per-thread.
+  std::thread::id tid;
+
+  bool operator==(const OnednnACLGemmKey& rhs) const {
+    return m == rhs.m && n == rhs.n && k == rhs.k && batch == rhs.batch &&
+           transpose_a == rhs.transpose_a && transpose_b == rhs.transpose_b &&
+           const_weight_ptr == rhs.const_weight_ptr && tid == rhs.tid;
+  }
+};
+
+std::thread::id getThreadId() {
+  static thread_local std::thread::id tid = std::this_thread::get_id();
+  return tid;
+};
+
+struct OnednnACLKeyHasher {
+  std::size_t operator()(const OnednnACLGemmKey& key) const {
+    std::size_t seed = std::hash<int>()(key.m);
+    hash_combine(seed, key.n);
+    hash_combine(seed, key.k);
+    hash_combine(seed, key.batch);
+    hash_combine(seed, key.transpose_a);
+    hash_combine(seed, key.transpose_b);
+    hash_combine(seed, key.const_weight_ptr);
+    hash_combine(seed, key.tid);
+    return seed;
+  }
+};
+
+using MatmulPrimitive = ideep::matmul_forward::super;
+using OnednnACLGemmCache =
+    std::unordered_map<OnednnACLGemmKey, std::shared_ptr<MatmulPrimitive>,
```

Comment:
Does each MatmulPrimitive hold a pre-packed weight? If yes, this cache will consume a huge amount of memory in dynamic shape cases with wide shape ranges, e.g. batch size from 1 to 1000.

Reply:
Yes, for sure. This is why such feature is not enabled by default. we eventually need a LRU cache for cached primitive.

Reply:
More challenging problem is primitive is not thread safe, leading to consuming more memory in multi-thread.

Reply:
OK, we should keep in mind this risk and be very careful in applying this feature on ARM cpus.

Reply:
I found there is a lru_cache implementation in ideep. I'll change the code to use this implementation.

Reply:
done